OEP 0028: Content theming in React

taranjeet opened the pull request:
This OEP provides guidelines for how to support content theming in React.

When the OEP is polished, a new PR will be opened against the [open-edx-proposals](https://github.com/edx/open-edx-proposals) repo.

taranjeet committed:
feat: Add OEP 28: Content Theming in React.

taranjeet committed:
chore: Fix grammatical bugs.

xitij2000 commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,206 @@
++-----------------+----------------------------------------------------------------+
+| Last Modified   | 2018-08-17                                                     |
++-----------------+----------------------------------------------------------------+
+| Authors         | Braden MacDonald                                               |
```
?? I think this should be your name @taranjeet

taranjeet replied:
Sure, I will update this.
I wrote Braden's name because he has authored the [gist](https://gist.github.com/bradenmacdonald/34bf8bfc8f737f9b23a2664091b8f115) originally.

bradenmacdonald replied:
You can put both ;)

taranjeet replied:
Sure @bradenmacdonald :) Will do that.

xitij2000 commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,206 @@
+This section describes the forces at play, including technological, political,
+social, and project local. These forces are probably in tension, and should
+be called out as such. The language in this section is value-neutral. It is
+simply describing facts.
```
I think this is something you can perhaps post in the PR description, but doesn't really make sense in the OEP itself, since that is a long-term document.

taranjeet replied:
This is the text from the [adr template](https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-templates/adr-based-template.rst#context). 
I will be updating this.

xitij2000 replied:
Ah! OK.

xitij2000 commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,206 @@
+
+> A component has a single responsibility when it has one reason to change.
+
+The above definition is unambiguous because there can be lots of ways to argue whether a component has more than one responsibility or not but it becomes easier to decide when this is conveyed in terms of change. The change can be cosmetic like color, spacing, animation, etc or functional like adding new feature etc.
```
I honestly don't think this is unambiguous. Can you give an example of such a component and the one reason it would change?

taranjeet replied:
As per my understanding a classic example of this can be `Button`.
A Button can be considered to have the following responsibilites (or change points)
* its color (styling)
* any icon (styling)
* its type or action (submit, normal) (functional)

Now if we have a single component Button, we can clearly see that it has more than resposibilities to change. Somebody can change its action, someone else can change the color and someone may add icon. Though these all is configurable through props, but you will end up writing a lot of logic in code, which will make the component difficult to update you in future. So in this case, if we were to apply Single Responsibility Principle, we would have the following components

* Button
* ColorButton
* IconButton
* ActionButton

xitij2000 replied:
You've just listed multiple "responsibilities" and multiple reasons to change for a simple Button. So what if I want to change add a button that has a different colour, an Icon (animated) and has an action? Do I then end up with AnimatedIconColourActionButton?

In the above case, how would you compose an ActionButton out of a regular Button if the regular button doesn't provide ways to change it? If the regular button allows you to specify the Action, why would you use an ActionButton?

taranjeet replied:
If we follow composition, we would end up with "AnimatedIconColourActionButton". 

There can be two ways to write such a button.
* One can be to write a single component which has support to do everything based on the props(and some conditional checks).
* The other can be to compose(composition here will actually require support via props and props.children).

To answer about how would you compose ActionButton, it won't be possible to if the original Button component doesn't have support.

My intent here was to highlight this that we should have components which are small and do only one thing. But I ended up giving a bad example. Sorry for the confusion.

But this should be kept in mind, that there will be times, where we will be in dilemma of either keeping a component as a single big component (with conditionals checks ) or creating a big component via composition of smaller components.

Commit:
Minor updates to match Hawthorn reality

taranjeet committed:
Update proposal with Decisions and Author name.

taranjeet committed:
Update draft with grammatical fixes.

taranjeet committed:
Add point about layout of data in redux store.

taranjeet committed:
Add Abstract and update Context.

taranjeet committed:
Add Consequences, Arbiter and Update Last modified.

taranjeet commented:
@xitij2000 : This is ready for your review. Please review. :)

nedbat committed:
Merge pull request #78 from edx/nedbat/hawthorn-10

Minor updates to match Hawthorn reality

bradenmacdonald commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,235 @@
+OEP-0028: Content Theming in React
+==================================
+
+.. This OEP template is based on Nygard's Architecture Decision Records.
```
Remove this line - it's only for the template, not OEPs based on the template.

bradenmacdonald commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,235 @@
+Context
+-------
+
+A modern UI system is composed of several big and customizable components. These components pose a challenge when updating the content/styling of any component. It becomes tougher when the change is required in the particular part of the UI, rather than the complete UI. This is one of the most common pains of the theming system.
```
I found this paragraph a bit confusing. Let's tweak the wording?

Suggestion: "Open edX instances frequently wish to customize not only the colors and styles of the Open edX UI, but also to add, remove, and modify significant parts of the UI to suit their use cases and branding. Doing this in a way that is understandable, clean, and easy to maintain can be a challenge unless the original UI is designed with this in mind."

bradenmacdonald commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,235 @@
+
+.. code-block:: js
+
+    class _Header extends React.PureComponent<Props, State> {
```
edX is actually not using TypeScript, so remove the `<Props, State>` part.

bradenmacdonald commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,235 @@
+Decision
+--------
+
+* We will build small modular components which will be non-customizable, stateless and contain HTML, CSS,  and certain logic. We can see an example where the header component is broken down into smaller components like SiteLogo, MainNav, and UserAvatar.
```
The `_Header` component you're showing here is something that I definitely meant to be customizable. So the example doesn't seem to match this description.

xitij2000 replied:
Additionally I don't think we are using the CSS in JS approach, so components should probably not contain CSS, Unless by component here we mean a component package, i.e. a js file, a test files and some styling in a folder = 1 component.

taranjeet replied:
@bradenmacdonald : I wanted to focus here on `SiteLogo`, `MainNav` and `UserAvatar` as non-customizable components but the `_Header` component came into limelight. Changing the example to be of `SiteLogo`.

@xitij2000 : Let us keep component simple(no component package) for now and hence I am removing css. 

Thanks for pointing it out :)

bradenmacdonald commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+* We will build the complete UI out of these small modular components. We will use composition to build a customizable component. These components will neither contain any HTML nor any logic. We can see in the above example where the _Header component is formed by the composition of SiteLogo, MainNav and UserAvatar component.
```
I find this a bit confusing, and the example below shows inheritance, not composition. (I think inheritance is the right approach).

bradenmacdonald commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,235 @@
+
+* We will provide support for a global redux store, which will act as a central place to store data. This store will access to all data available to the system.
+
+* We will consider the layout of the data in the global store as an API. We will provide support to pre-fill the store with some common data like current user, current course, list of courses enrolled, etc. and add the less commonly needed data from REST API's using built-in redux actions. We will announce breaking changes if the layout of the data changes in store.
```
I feel like the Open edX global redux store should be considered a stable API, but if themes want to read/write data that's not already part of that store, they should store it in their own separate redux store, and do their own fetching of data from REST APIs etc., so as not to conflict with the Open edX redux store, other themes, other plugins, etc. Thoughts?

xitij2000 replied:
I don't think perhaps we can just namespace the stable vs unstable events and state, and namspace each plugin/theme. I also feel that instead of making the state itself stable API, the stable bit can be small selector methods that dig through the state to get the data you want. So you would have a function like:

```getUserProfileInfo = (state, userid) => state.users[userid].profile```

If the state shape changes, the worst case would be that this function becomes more complex and now pulls in data from multiple sources but still provides the same API.

taranjeet replied:
@xitij2000 @bradenmacdonald : From the discussion, there can be two approaches
* State as a stable API 
* OR State consisting of selector method as a stable API

I think we can keep Braden's point of considering the state as a stable API and any theme can create their own separate redux store if they are requesting some data which is not present in the Open edx redux store. In the [gist](https://gist.github.com/bradenmacdonald/34bf8bfc8f737f9b23a2664091b8f115) Matjaz also [mentioned](https://gist.github.com/bradenmacdonald/34bf8bfc8f737f9b23a2664091b8f115#gistcomment-2624625) about keeping the state as API.
Let me know if I am missing anything here?

bradenmacdonald commented on oeps/oep-0028-content-theming-in-react.rst:
```diff
@@ -0,0 +1,235 @@
+* We will consider the layout of the data in the global store as an API. We will provide support to pre-fill the store with some common data like current user, current course, list of courses enrolled, etc. and add the less commonly needed data from REST API's using built-in redux actions. We will announce breaking changes if the layout of the data changes in store.
+
+
+* We will use containers to access data from the redux store and provide it to components via props.
```
Can you elaborate on this? I don't think this is enough of an explanation. Containers are not a standard React concept, right?

xitij2000 replied:
Yup, I don't think it's enough to say containers. Containers is a name people came up with for certain kinds of non-pure components that interact with external data like the redux store. It would be better to say that we will separate the redux data access from the component functionality, such that you can import the non-redux-connected component as well as a redux-connected version of the same component.

taranjeet committed:
Add author and remove template comment.

taranjeet committed:
Update Context and remove typescript support.

taranjeet committed:
Update small components point under Decision.

taranjeet committed:
Update Customizable component point and example.

taranjeet committed:
Explain container in more detail.

taranjeet committed:
explain about global redux store.

taranjeet commented:
@xitij2000 @bradenmacdonald I have addressed all the review comments. Hence I am moving this PR to [upstream edx](https://github.com/edx/open-edx-proposals) repo as [PR 79](https://github.com/edx/open-edx-proposals/pull/79). I will keep this PR and if required will post the discussion from here on the upstream PR.

I will be updating my branch name to `opencraft/taranjeet/content-theming-react`.
